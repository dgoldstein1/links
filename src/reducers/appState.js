import { SET_LOADING, UPDATE_VIEW, SET_FATAL_ERROR } from "../actions/appState";

const initialState = {
  loading: true, // the app is / isn't loading
  fatalError: undefined, // error which  causes app to not load
  view: "splash" // one of "explore", "path", "splash", "about", "settings"
};

const visitReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_FATAL_ERROR:
      return Object.assign({}, state, {
        fatalError: action.fatalError
      });
    // set when the app is loading
    case SET_LOADING:
      return Object.assign({}, state, {
        loading: action.loading === undefined ? false : action.loading
      });
    // update where the user is currently looking
    case UPDATE_VIEW:
      return Object.assign({}, state, {
        view: action.view
      });
    default:
      return state;
  }
};

export default visitReducer;
