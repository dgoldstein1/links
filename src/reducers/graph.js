import _ from "lodash";

const initialState = {
  rootNode: {}, // root node of graph (what is in search or 'start from')
  selectedNode: {
    // node currently in view
    loading: true,
    error: "",
    description: "",
    img: "",
    node: {}
  },
  targetNode: {}, // target node in path ('ending at..')
  graph: {
    nodes: [],
    edges: []
  },
  maxShortestPaths: 5,
  directedShortestPath: true,
  loading: false,
  error: undefined,
  layout: "hierarchy",
  maxNeighbors: 15,
  topInfo: {
    betweenessEdges: [],
    betweenessNodes: [],
    pageRank: []
  }
};

const graphReducer = (state = initialState, action) => {
  switch (action.type) {
    case "CLEAR_GRAPH":
      return Object.assign({}, state, {
        graph: {
          nodes: [],
          edges: []
        }
      });
    case "SET_DIRECTED_SHORTEST_PATH":
      return Object.assign({}, state, {
        directedShortestPath: action.directedShortestPath
      });
    case "SET_MAX_NEIGHBORS":
      return Object.assign({}, state, {
        maxNeighbors: action.maxNeighbors
      });
    case "SET_GRAPH_ERROR":
      return Object.assign({}, state, {
        error: action.error
      });
    case "SET_GRAPH_LAYOUT":
      return Object.assign({}, state, {
        layout: action.layout
      });
    case "SET_GRAPH_LOADING":
      return Object.assign({}, state, {
        loading: action.loading
      });
    case "SET_ROOT_NODE":
      return Object.assign({}, state, {
        rootNode: _generateRoot(action.node.label, action.node.id)
      });
    case "SET_MAX_SHORTEST_PATHS":
      return Object.assign({}, state, {
        maxShortestPaths: action.maxShortestPaths
      });
    case "SET_ALL_PATHS_UNIQUE":
      return Object.assign({}, state, {
        pathsAreUnique: action.pathsAreUnique
      });

    case "SET_SELECTED_NODE":
      return Object.assign({}, state, {
        selectedNode: {
          ...state.selectedNode,
          node: action.node
        }
      });
    case "SET_SELECTED_NODE_INFO":
      return Object.assign({}, state, {
        selectedNode: {
          ...state.selectedNode,
          description: action.description,
          img: action.img,
          loading: action.loading,
          error: action.error,
          centrality: action.centrality
        }
      });
    case "SET_TARGET_NODE":
      return Object.assign({}, state, {
        targetNode: action.node
      });

    case "SET_GRAPH_PATH":
      // create array of all nodes from entries
      let nodes = action.entries.map(e => ({
        id: e.value,
        label: e.key,
        size: 1,
        color: "red"
      }));
      // create edges from each path in graph
      let edges = [];
      let entriesReverseMap = {};
      action.entries.forEach(e => (entriesReverseMap[e.value] = e.key));
      action.paths.forEach(p => {
        p.forEach((id, i) => {
          // dont add edge if we've reached the end
          if (i === p.length - 1) return;
          // lookup this entry and next entry
          edges.push({
            id: `${id}->-->--${p[i + 1]}`,
            source: id,
            target: p[i + 1],
            color: "red"
          });
        });
      });
      return Object.assign({}, state, {
        graph: { nodes, edges }
      });
    case "ADD_NEIGHBORS_TO_GRAPH":
      // set initial position to source
      action.neighbors = action.neighbors.map(n => ({
        ...n,
        size: 1
      }));
      // get list of edges to add
      let edgesToAdd = [];
      action.neighbors.forEach(n => {
        if (!n.id) console.log(n);
        edgesToAdd.push({
          id: `${action.node.id}->-->--${n.id}`,
          source: action.node.id,
          target: n.id,
          color: "black"
        });
      });
      // add in node to make sure it's in list of nodes
      action.neighbors.push(action.node);
      // return state
      return Object.assign({}, state, {
        graph: {
          nodes: _.unionBy(state.graph.nodes, action.neighbors, "id"),
          edges: _.unionBy(state.graph.edges, edgesToAdd, "id")
        }
      });
    default:
      return state;
  }
};

export function _generateRoot(label, id) {
  return {
    label,
    id,
    x: 0,
    y: 0,
    size: 1
  };
}

export default graphReducer;
