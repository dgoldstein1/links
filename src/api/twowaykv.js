import { makeRequest } from "./utils";
import { store } from "../reducers";

/**
 * fetches random nodes from twowaykv
 * @param n, number of random items to fetch
 * @return:
 *{
 *  success : bool,
 *   data : [{key, value}],
 *   error : string
 *}
 **/
export function random(n) {
  return makeRequest({
    method: "get",
    url: `${store.getState().appState.graphConfig.kvEndpoint}/random?n=${n}`,
    onErrorPrefix: "Error getting random entry"
  });
}

/**
 * gets entries from IDs
 * @param ids []int
 * @return same response type as above
 **/
export function entriesFromValues(ids) {
  return makeRequest({
    method: "post",
    url: `${
      store.getState().appState.graphConfig.kvEndpoint
    }/entriesFromValues`,
    onErrorPrefix: "Error getting entries from values",
    body: ids
  });
}

/**
 * searches 'startswith' on keys
 * @param s {string}
 * @return same response type as above
 **/
export function search(s) {
  return makeRequest({
    method: "get",
    url: `${store.getState().appState.graphConfig.kvEndpoint}/search?q=${s}`,
    onErrorPrefix: "Error searching keys"
  });
}
