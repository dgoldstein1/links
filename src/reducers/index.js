// redux utils
import { combineReducers, createStore, applyMiddleware } from "redux";
import { createLogger } from "redux-logger";

// reducers in this directory
import appState from "./appState";
import graph from "./graph";

const Reducer = combineReducers({
  appState,
  graph,
});

const logger = createLogger({
  collapsed: true,
  diff: true,
});

let middleware = {};
if (process.env.REACT_APP_ENV === "development") {
  middleware = applyMiddleware(logger);
}

const store = createStore(Reducer, middleware);

export { store };
