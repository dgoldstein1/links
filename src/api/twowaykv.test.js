import { random, entriesFromValues, search } from "./twowaykv";
import moxios from "moxios";
import { store } from "../reducers";
import sinon from "sinon";
import axios from "axios";

describe("twowaykv", () => {
  beforeEach(() => {
    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
  });

  describe("random()", () => {
    beforeEach(() => {
      moxios.install();
    });

    afterEach(() => {
      moxios.uninstall();
    });
    let testTable = [
      {
        name: "successful request",
        n: 1,
        url: `/services/twowaykv/random?n=1`,
        responseCode: 200,
        data: [
          {
            key: "key",
            value: 234234
          }
        ]
      }
    ];
    testTable.forEach(t => {
      it(t.name, done => {
        let onFulfilled = sinon.spy();
        moxios.stubRequest(t.url, {
          status: t.responseCode,
          response: t.data
        });
        random(t.n).then(r => {
          expect(r).toMatchSnapshot();
          done();
        });
      });
    });
  });
  describe("entriesFromValues", () => {
    beforeEach(() => {
      moxios.install();
    });

    afterEach(() => {
      moxios.uninstall();
    });
    let testTable = [
      {
        name: "success",
        ids: [234234],
        url: `/services/twowaykv/entriesFromValues`,
        responseCode: 200,
        data: [
          {
            key: "key",
            value: 234234
          }
        ]
      }
    ];
    testTable.forEach(t => {
      it(t.name, done => {
        let onFulfilled = sinon.spy();
        moxios.stubRequest(t.url, {
          status: t.responseCode,
          response: t.data
        });
        entriesFromValues(t.ids).then(r => {
          expect(r).toMatchSnapshot();
          done();
        });
      });
    });
  });
  describe("search", () => {
    beforeEach(() => {
      moxios.install();
    });

    afterEach(() => {
      moxios.uninstall();
    });
    let testTable = [
      {
        name: "success",
        s: "test",
        url: `/services/twowaykv/search?q=test`,
        responseCode: 200,
        data: [
          {
            key: "key",
            value: 234234
          }
        ]
      }
    ];
    testTable.forEach(t => {
      it(t.name, done => {
        let onFulfilled = sinon.spy();
        moxios.stubRequest(t.url, {
          status: t.responseCode,
          response: t.data
        });
        search(t.s).then(r => {
          expect(r).toMatchSnapshot();
          done();
        });
      });
    });
  });
});
