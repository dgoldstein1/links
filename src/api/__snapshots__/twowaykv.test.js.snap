// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`twowaykv entriesFromValues success 1`] = `
Object {
  "data": Array [
    Object {
      "key": "key",
      "value": 234234,
    },
  ],
  "success": true,
}
`;

exports[`twowaykv random() successful request 1`] = `
Object {
  "data": Array [
    Object {
      "key": "key",
      "value": 234234,
    },
  ],
  "success": true,
}
`;

exports[`twowaykv search success 1`] = `
Object {
  "data": Array [
    Object {
      "key": "key",
      "value": 234234,
    },
  ],
  "success": true,
}
`;
